File: 2025-08-August-GFG-POTD/August 23 - Allocate Minimum Pages/Solution.java
class Solution {
    public int findPages(int[] arr, int k) {
        int n = arr.length;
        
        // Edge case: more students than books
        if (k > n) return -1;
        
        int low = 0, high = 0;
        
        // Low = max book, High = sum of all books
        for (int pages : arr) {
            low = Math.max(low, pages);
            high += pages;
        }
        
        int result = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            
            if (isPossible(arr, n, k, mid)) {
                result = mid;
                high = mid - 1;  // try for a smaller maximum
            } else {
                low = mid + 1;   // need a larger maximum
            }
        }
        return result;
    }
    
    private boolean isPossible(int[] arr, int n, int k, int maxPages) {
        int students = 1;
        int pagesSum = 0;
        
        for (int pages : arr) {
            if (pagesSum + pages > maxPages) {
                students++;
                pagesSum = pages;
                
                if (students > k) return false;
            } else {
                pagesSum += pages;
            }
        }
        return true;
    }
}

File: 2025-08-August-GFG-POTD/August 23 - Allocate Minimum Pages/Solution.py
class Solution:
    def findPages(self, arr, k):
        n = len(arr)
        
        # Edge case: more students than books
        if k > n:
            return -1
        
        low = max(arr)   # minimum limit
        high = sum(arr)  # maximum limit
        
        result = -1
        while low <= high:
            mid = (low + high) // 2
            
            if self.isPossible(arr, n, k, mid):
                result = mid
                high = mid - 1  # try smaller maximum
            else:
                low = mid + 1   # need larger maximum
                
        return result
    
    def isPossible(self, arr, n, k, maxPages):
        students = 1
        pagesSum = 0
        
        for pages in arr:
            if pagesSum + pages > maxPages:
                students += 1
                pagesSum = pages
                
                if students > k:
                    return False
            else:
                pagesSum += pages
        return True
